### ISSUE TITLE
location 'replace' not working

### ISSUE THREAD

**Issue by somentelucas** [angular-ui/ui-router#1007]
I'm doing a transition to the same state (different params) and I intend to NOT keep this transition in history, so I did
```
$state.go("state.name", {param: param}, {location: 'replace'});
```
But when I go back it still goes through all the previous states.

**Comment by danharper** [comment]
For anyone using Ionic, there's a separate issue with their Back button. There's a couple of quite hacks in this issue:

https://github.com/driftyco/ionic/issues/1287

**Comment by alexstanbury** [comment]
I agree with @OverZealous, there really needs to be a way to access the options passed to ui-sref-opts, I am currently stuck as I need to interrupt a state change and then continue later on, but there is no way for me to find out whether I passed any options with the state change.

**Comment by christopherthielen** [comment]
I'll accept a PR that adds options as param #5 to state events

**Comment by mkoczorowski** [comment]
@OverZealous Thanks!
adding data.permissions: false worked for me

**Comment by petetaxi-test** [comment]
I was getting the same error as @FreakTheMighty plunkr. 

Found a simple, albeit hacky workaround - put a slight delay on the $state.go using $timeout e.g. 10ms so that it happens after the first state transition completes.

**Comment by nateabele** [comment]
`options` has been added to `$stateChangeStart`. Can we close this?

**Comment by stecd** [comment]
@nateabele and how do you solve this issue using `options` from `$statechangeStart`?

**Comment by johncant** [comment]
Sorry for adding noise, but $state.go("state.name", {param: param}, {location: 'replace'}); is replacing the current state with the new state for me, which is what I expected, so the back button then goes from the new state to the previous state, skipping the current state. Recommend closing, since there's been no plunkr/jsfiddle in over 18 months, and two people have eliminated angular-ui-router as the cause of their manifestation of this issue. :)

**Comment by Mirodil** [comment]
still not wokring +1

**Comment by jpost** [comment]
Not working for me either with v1.2.4-nightly-1917 +1

**Comment by keshavwork** [comment]
still not wokring +1

**Comment by doorty** [comment]
still not working ++

**Comment by fijimunkii** [comment]
`$state.go('place.to.be', $stateParams, { location: 'replace' })` is working for me.

I think the issue most people run into here is passing `{ location: 'replace' }` as the second argument instead of the third.

**Comment by abachuk** [comment]
still not working +++

**Comment by christopherthielen** [comment]
@keshavwork 
> still not wokring +1
@doorty 
> still not working ++
@abachuk 
> still not working +++


These comments are not helpful.  Post a plunker demonstrating that it's not working.  http://bit.ly/UIR-Plunk

As far as I can tell, [`{location: 'replace'}` works like it's supposed to](http://embed.plnkr.co/C9smZuPGXo9P0jS9N2Fo/?show=preview). Prove me wrong by posting a reproduction.

**Comment by inad9300** [comment]
Still not working ++++

I appreciate the effort you put on developing this open source tool, but please, if many people say this feature does not work, at least take the time to verify it properly, otherwise who is going to rely on it?

I modified your plunkr to show the case when it doesn't work: when you programmatically call `go()` with `{location: 'replace'}`, which I think is the most common use case (as oppose to embed the options on the HTML).

There you can see how states are pushed into the history despite the "replace" intention, and you can still go back to them: https://plnkr.co/edit/i3R8jaRS6TrMz3KKHT10?p=preview

**Comment by dkordik** [comment]
@inad9300 your example seems to work correctly for me (and your example is pointing to angular-ui-router@latest, so maybe it was recently fixed in "latest"?)

**Comment by dkordik** [comment]
@inad9300 here's an updated plnkr with the state.go pieces split out from the template, and using `{ location: 'replace' }` instead of `replace: true`.

Seems to work for both the buttons and the select box examples.
https://plnkr.co/edit/i3R8jaRS6TrMz3KKHT10?p=preview

**Comment by inad9300** [comment]
My mistake, sorry. So apparently it works after all :) Thanks for taking the time!

**Comment by christopherthielen** [comment]
@inad9300 next time please be considerate of the amount of time OSS maintainers spend.   Snarky comments like "Still not working ++++" make me reconsider my involvement in OSS.

**Comment by evirca** [comment]
I had the same issue and after research i have realized i was wrongly understanding the concept of replace.
**When using location.replace you are replacing (exclude from history) the current page (not the one navigating to) with the page you are navigation to**.  So in ui-router case, when using `$state.go('place.to.be', $stateParams, { location: 'replace' })` it will actually replace the current state with the state 'place.to.be'. 
In other words, if you want to navigate to a page R that just redirects to another page X make sure you are navigation from page R () by using `$state.go('X', $stateParams, { location: 'replace' })` . This way R will be replaced with page X

**Comment by juliobguedes** [comment]
Has this problem been solved? I'm using ui-router v0.2.13 along another group of dependencies (which I unfortunately cannot post here) and I'm facing the problem :cry: 

What I have here is: I have a list of objects, and clicking in an object on the list redirects me to an intermediary screen with a `state.go()`. In this intermediary screen, I use `location: 'replace'` to replace with the final screen after some logic.

When I click the navigator button once it works and I go to the previous screen, but clicking twice keeps me on the previous screen after a redirecting...

My exact code is `$state.go(finalScreen, finalParams, { location: 'replace' });`
